### PR TITLE
Add span logFields: wait_goroutine_capacity_time_ms and min_weighted_…

### LINKED ISSUE #5316: Add  wait time created by the LimitedRountTripper as a span log.
Currently a query with too high parallelism will get limited by the limitedRoundtripper https://github.com/grafana/loki/blob/main/pkg/querier/queryrange/limits.go#L292

However there's no trace of this happening, it would be great to add a span log and log that includes the time waited for goroutine capacity.

### PR DESCRIPTION
**Pull request by niuguy**
…parallism

**What this PR does / why we need it**:
https://github.com/grafana/loki/issues/5316

**Which issue(s) this PR fixes**:
Fixes #5316 

**Special notes for your reviewer**:
Please review and if possible give some hints on how to test this

**Checklist**
- [x ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: pkg/querier/queryrange/limits.go [hunk 1]
```diff
@@ -509,6 +513,8 @@ func (rt limitedRoundTripper) RoundTrip(r *http.Request) (*http.Response, error)
 
 	sem := semaphore.NewWeighted(int64(parallelism))
 
+	start := time.Now()
```
Comment (dannykopping) [inline pkg/querier/queryrange/limits.go:516]: This needs to move before the semaphore is acquired on line 527. The handler is called multiple times, so we can't create a reference point beforehand because it'll seem like all subsequent requests are taking longer and longer.

**Review by dannykopping — COMMENTED** [review]
Thank you for the contribution @niuguy!

Have you tested this out? I'd be interested to see an example trace where these values are calculated.

File: pkg/querier/queryrange/limits.go (hunk 1 above)
Comment (niuguy) [reply pkg/querier/queryrange/limits.go]: @dannykopping  please check the latest two commits where I wrapped the Aquire and the timing stuff in one method. It's primarily for the benefit of test.  There should be some elegant approach to do this , please feel free to point it out. Thanks

File: pkg/querier/queryrange/limits.go [hunk 2]
```diff
@@ -461,9 +485,12 @@ func (rt limitedRoundTripper) Do(c context.Context, request queryrangebase.Reque
 		cancel()
 	}()
 
-	if span := opentracing.SpanFromContext(ctx); span != nil {
+	span := opentracing.SpanFromContext(ctx)
+
+	if span != nil {
 		request.LogToSpan(span)
 	}
+
```
Comment (dannykopping) [inline pkg/querier/queryrange/limits.go:493]: What's the purpose of this change?

File: pkg/querier/queryrange/limits_test.go [hunk 3]
```diff
@@ -623,3 +623,41 @@ func Test_MaxQuerySize_MaxLookBackPeriod(t *testing.T) {
...
+	go tryAcquire(1)
+	go tryAcquire(1)
+
+	// Collect waiting times
+	waiting1 := <-waitingTimes
+	waiting2 := <-waitingTimes
+	waiting3 := <-waitingTimes
+
+	// Check that at least one request waited for some time
+	if waiting1 == 0 && waiting2 == 0 && waiting3 == 0 {
```
Comment (dannykopping) [inline pkg/querier/queryrange/limits_test.go:660]: The comment and this assertion don't make sense to me.
If all 3 are waiting for 0ms then I don't understand what you're testing.

File: pkg/querier/queryrange/limits.go [hunk 4]
```diff
@@ -453,6 +454,29 @@ func NewLimitedRoundTripper(next queryrangebase.Handler, limits Limits, configs
 	return transport
 }
 
+type SemaphoreWithTiming struct {
+	sem *semaphore.Weighted
+}
+
+func NewSemaphoreWithTiming(max int64) *SemaphoreWithTiming {
+	return &SemaphoreWithTiming{
+		sem: semaphore.NewWeighted(int64(max)), // Replace with your semaphore creation logic
```
Comment (dannykopping) [inline pkg/querier/queryrange/limits.go:463]:
```suggestion
		sem: semaphore.NewWeighted(int64(max)),
```

File: pkg/querier/queryrange/limits.go [hunk 5]
```diff
@@ -453,6 +454,29 @@ func NewLimitedRoundTripper(next queryrangebase.Handler, limits Limits, configs
...
+func (s *SemaphoreWithTiming) Acquire(ctx context.Context, n int64) (int64, error) {
+	start := time.Now()
+
+	if err := s.sem.Acquire(ctx, int64(n)); err != nil {
+		return 0, err
+	}
+
+	elapsed := time.Since(start)
+
+	return elapsed.Milliseconds(), nil
```
Comment (dannykopping) [inline pkg/querier/queryrange/limits.go:477]: I would just return `time.Since(start)` as the value here, and call `String()` on it when you log it so it'll show a useful value at any timescale.

**Review by dannykopping — COMMENTED** [review]
@niuguy I've taken another look. This is looking more promising, but there are a couple things in the implementation I raised.

File: pkg/querier/queryrange/limits.go (hunk 2 above)
Comment (niuguy) [reply pkg/querier/queryrange/limits.go:493]: I'd like to move this "span" out so it can be reused below in the handler

File: pkg/querier/queryrange/limits.go (hunk 4 above)
Comment (niuguy) [reply pkg/querier/queryrange/limits.go:463]: 😅

File: pkg/querier/queryrange/limits_test.go (hunk 3 above)
Comment (niuguy) [reply pkg/querier/queryrange/limits_test.go]: Please check. I updated the test which simulate running two requests at full capacity , when the 3rd comes this method should expect waiting time >0 returned

File: pkg/querier/queryrange/limits.go [hunk 6]
```diff
@@ -493,10 +518,20 @@ func (rt limitedRoundTripper) Do(c context.Context, request queryrangebase.Reque
...
+
+			if err != nil {
 				return nil, fmt.Errorf("could not acquire work: %w", err)
 			}
-			defer sem.Release(int64(1))
+
+			if span != nil {
+				span.LogFields(
+					otlog.String("wait_goroutine_capacity_time", elapsed.String()),
+					otlog.Int64("min_weighted_parallelism", int64(parallelism)),
```
Comment (dannykopping) [inline pkg/querier/queryrange/limits.go:530]:
```suggestion
					otlog.String("wait_time", elapsed.String()),
					otlog.Int64("max_parallelism", int64(parallelism)),
```

File: pkg/querier/queryrange/limits_test.go [hunk 7]
```diff
@@ -623,3 +623,61 @@ func Test_MaxQuerySize_MaxLookBackPeriod(t *testing.T) {
...
+	// Find and check the waiting time for the third goroutine
+	var waiting3 int64
+	for _, waiting := range waitingDurations {
+		if waiting.GoroutineID == 3 {
+			waiting3 = waiting.WaitingTime
+			break
+		}
+	}
+
+	// Check that the waiting time for the third request is at least 10 milliseconds
```
Comment (dannykopping) [inline pkg/querier/queryrange/limits_test.go:678]: If you're expecting it to be at least 10ms, express that in the test please

Comment (dannykopping) [reply pkg/querier/queryrange/limits_test.go:678]: You could use `require.GreaterOrEqual(t, waiting3, 10*time.Millisecond)`

**Review by dannykopping — COMMENTED** [review]
Looking good! Last couple small comments.

File: pkg/querier/queryrange/limits_test.go (hunk 7 above)
Comment (niuguy) [reply pkg/querier/queryrange/limits_test.go:678]: sorry this is a typo  should be larger than 0 milliseconds, actually it should be  waiting3>0  && waiting3 <=5

Comment (niuguy) [reply pkg/querier/queryrange/limits_test.go]: have updated

**Comment by dannykopping** [comment]
Alright @niuguy, now to fight CI and then we can get this merged 🥲

**Comment by niuguy** [comment]
Hey @dannykopping  the CI has been pending for a while would you be able to have a look

**Review by dannykopping — APPROVED** [review]
Thanks very much for your contribution and your patience, @niuguy!
This will certainly be useful.

**Comment by niuguy** [comment]
Thank you a lot  for your guidance @dannykopping ,  happy to contribute!
